signingworker/worker.py

```diff
@@ -0,0 +1,187 @@
+    def get_manifest(self, url):
+        r = requests.get(url)
+        r.raise_for_status()
+        return r.json()
+
+    def sign(self, task_id, run_id, task):
+        payload = task["payload"]
+        manifest_url = payload["signingManifest"]
+        signing_manifest = self.get_manifest(manifest_url)
+        # TODO: better way to extract filename
+        url_prefix = "/".join(manifest_url.split("/")[:-1])
+        for e in signing_manifest:
+            # TODO: "mar" is too specific, change the manifest
+            file_url = "{}/{}".format(url_prefix, e["mar"])
+            abs_filename = self.download_and_sign_file(
+                task_id, run_id, file_url, e["hash"])
+            # Update manifest data with new values
+            e["hash"] = get_hash(abs_filename)
+            e["size"] = os.path.getsize(abs_filename)
+        _, manifest_file = tempfile.mkstemp()
+        with open(manifest_file, "wb") as f:
+            json.dump(signing_manifest, f, indent=2, sort_keys=True)
+        log.debug("Uploading manifest for t: %s, r: %s", task_id, run_id)
+        self.create_artifact(task_id, run_id, "public/env/manifest.json",
+                             manifest_file, "application/json")
+
+    def download_and_sign_file(self, task_id, run_id, url, checksum):
+        work_dir = tempfile.mkdtemp()
+        # TODO: better parsing
+        filename = urlparse.urlsplit(url).path.split("/")[-1]
+        abs_filename = os.path.join(work_dir, filename)
+        r = requests.get(url)
+        r.raise_for_status()
+        with open(abs_filename, 'wb') as fd:
+            for chunk in r.iter_content(4096):
+                fd.write(chunk)
+        got_checksum = get_hash(abs_filename)
+        if not got_checksum == checksum:
+            log.debug("Checksum mismatch, cleaning up...")
+            shutil.rmtree(work_dir)
+            raise ChecksumMismatchError("Expected {}, got {} for {}".format(
+                checksum, got_checksum, url
+            ))
+        self.sign_file(work_dir, filename)
+        self.create_artifact(task_id, run_id, "public/env/%s" % filename,
+                             abs_filename)
+        return abs_filename
+
+    def create_artifact(self, task_id, run_id, dest, abs_filename,
+                        content_type="application/octet-stream"):
+        log.debug("Uploading artifact %s (t: %s, r: %s) from %s (%s)", dest,
+                  task_id, run_id, abs_filename, content_type)
+        # TODO: better expires
+        res = self.tc_queue.createArtifact(
+            task_id, run_id, dest,
+            {
+                "storageType": "s3",
+                "contentType": content_type,
+                "expires": arrow.now().replace(weeks=2).isoformat()
+            }
+        )
+        log.debug("Got %s", res)
+        put_url = res["putUrl"]
+        log.debug("Uploading to %s", put_url)
+        taskcluster.utils.putFile(abs_filename, put_url, content_type)
```

Comment:
Is this something you should be retrying? It's probably one of the more likely things to fail since it's going to be a bigger operation than most.